File: leetcode/RemoveElement.java
package leetcode;

import java.util.Arrays;

/**
 * This class removes an element from a given array in place."
 */

public class RemoveElement {

    // Time: O(n)
    // Space: O(1)
    // removes element from given array in place using two pointers
    public static int removeElement(int[] nums, int val) {
        // initialize startIndex to 0
        int startIndex = 0;
        // initialize endIndex to 0
        int endIndex = 0;
        // while startIndex is less than nums length
        while (startIndex < nums.length) {
            // if nums at startIndex is equal to val
            if (nums[startIndex] == val) {
                // increase startIndex by 1
                startIndex++;
            } else {
                // else set nums at endIndex equal to nums at startIndex
                nums[endIndex] = nums[startIndex];
                // increase startIndex by 1
                startIndex++;
                // increase endIndex by 1
                endIndex++;
            }
        }
        // after loop return endIndex
        System.out.println(Arrays.toString(nums));
        return endIndex;
    }

    // main method
    public static void main(String[] args) {
        int[] nums = {3,2,2,3};
        System.out.println(removeElement(nums, 3));
    }

}

File: leetcode/python/remove_element.py
# Given an integer array nums and an integer val, remove all occurrences of val in nums in-place. 
# The order of the elements may be changed. Then return the number of elements in nums which are not equal to val.
# Consider the number of elements in nums which are not equal to val be k, to get accepted, 
# you need to do the following things:

#     Change the array nums such that the first k elements of nums contain the elements which are not equal to val. The remaining elements of nums are not important as well as the size of nums.
#     Return k.

# Example 1:

# Input: nums = [3,2,2,3], val = 3
# Output: 2, nums = [2,2,_,_]

def removeElement(nums: list[int], val: int) -> int:
    # init write
    write = 0

    # for each read in range of length of nums
    for read in range(len(nums)):
        # if nums at read is not equal to val
        if nums[read] != val:
            # set nums at write to nums at read
            nums[write] = nums[read]
            # increase write by 1
            write += 1
    
    # return write
    return write

print(removeElement([3,2,2,3], 3))

# time: O(n)
# space: O(1)
# pattern: two pointers